From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DeeChael <DeeChael@outlook.com>
Date: Sun, 27 Aug 2023 22:40:02 +0800
Subject: [PATCH] remove useless api


diff --git a/src/main/kotlin/net/deechael/elements/api/ElementManager.kt b/src/main/kotlin/net/deechael/elements/api/ElementManager.kt
index b9f542d92b15b568218f15d814087e3f99917e2c..b6b42afd064d3217cc386732af4f1cb4b6a438c8 100644
--- a/src/main/kotlin/net/deechael/elements/api/ElementManager.kt
+++ b/src/main/kotlin/net/deechael/elements/api/ElementManager.kt
@@ -1,12 +1,6 @@
 package net.deechael.elements.api
 
-import net.deechael.elements.api.application.ApplicationTrigger
-import net.deechael.elements.api.application.ElementApplicationManager
-import net.deechael.elements.api.application.source.SourceManager
-import net.deechael.elements.api.mute.ElementMuteManager
 import net.deechael.elements.api.reaction.ElementReaction
-import net.deechael.elements.api.reaction.ElementReactionTrigger
-import java.awt.Color
 
 interface ElementManager {
 
@@ -16,16 +10,6 @@ interface ElementManager {
 
     fun getElementTypes(): List<ElementType>
 
-    fun registerElementType(
-        id: String,
-        icon: String,
-        color: Color,
-        applicable: Boolean,
-        reactable: Boolean,
-        isReactionElement: Boolean = false,
-        trigger: ApplicationTrigger = ApplicationTrigger.EMPTY_TRIGGER
-    ): ElementType
-
     fun hasElementReaction(id: String): Boolean
 
     fun getElementReaction(id: String): ElementReaction
@@ -34,18 +18,4 @@ interface ElementManager {
 
     fun getElementReactionsWithFormerType(formerElementType: ElementType): List<ElementReaction>
 
-    fun registerElementReaction(
-        id: String,
-        color: Color,
-        formerElementType: ElementType,
-        latterElementTypes: Array<ElementType>,
-        trigger: ElementReactionTrigger
-    ): ElementReaction
-
-    fun getApplicationManager(): ElementApplicationManager
-
-    fun getSourceManager(): SourceManager
-
-    fun getMuteManager(): ElementMuteManager
-
 }
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/elements/api/application/ElementApplicationManager.kt b/src/main/kotlin/net/deechael/elements/api/application/ElementApplicationManager.kt
deleted file mode 100644
index a309c629f6d07942774e7c9a40f757e41016dc00..0000000000000000000000000000000000000000
--- a/src/main/kotlin/net/deechael/elements/api/application/ElementApplicationManager.kt
+++ /dev/null
@@ -1,14 +0,0 @@
-package net.deechael.elements.api.application
-
-import net.deechael.elements.api.application.source.ApplicationSource
-import org.bukkit.entity.Entity
-
-interface ElementApplicationManager {
-
-    fun hasApplied(entity: Entity): Boolean
-
-    fun getApplication(entity: Entity): ElementApplication
-
-    fun doElementalDamage(sufferer: Entity, source: ApplicationSource, damage: Double)
-
-}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/elements/api/application/source/SourceManager.kt b/src/main/kotlin/net/deechael/elements/api/application/source/SourceManager.kt
deleted file mode 100644
index a12534301abc87171a7982cd530a9b5555866fe1..0000000000000000000000000000000000000000
--- a/src/main/kotlin/net/deechael/elements/api/application/source/SourceManager.kt
+++ /dev/null
@@ -1,36 +0,0 @@
-package net.deechael.elements.api.application.source
-
-import net.deechael.elements.api.ElementGauge
-import net.deechael.elements.api.ElementType
-import net.deechael.elements.api.reaction.ElementReaction
-import org.bukkit.Location
-import org.bukkit.entity.Entity
-import org.bukkit.plugin.Plugin
-
-interface SourceManager {
-
-    fun entity(
-        entity: Entity,
-        elementType: ElementType,
-        gauge: ElementGauge
-    ): EntitySource
-
-    fun environment(
-        location: Location,
-        elementType: ElementType,
-        gauge: ElementGauge
-    ): EnvironmentSource
-
-    fun plugin(
-        plugin: Plugin,
-        elementType: ElementType,
-        gauge: ElementGauge
-    ): PluginSource
-
-    fun reaction(
-        reaction: ElementReaction,
-        elementType: ElementType,
-        gauge: ElementGauge
-    ): ReactionSource
-
-}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/elements/api/mute/ElementMuteManager.kt b/src/main/kotlin/net/deechael/elements/api/mute/ElementMuteManager.kt
deleted file mode 100644
index e45a24a749bec9cf42d5190b97a8fb419dd6fc2e..0000000000000000000000000000000000000000
--- a/src/main/kotlin/net/deechael/elements/api/mute/ElementMuteManager.kt
+++ /dev/null
@@ -1,16 +0,0 @@
-package net.deechael.elements.api.mute
-
-import net.deechael.elements.api.ElementType
-import org.bukkit.entity.Entity
-
-interface ElementMuteManager {
-
-    fun addMuteElementType(entity: Entity, elementType: ElementType)
-
-    fun removeMuteElementType(entity: Entity, elementType: ElementType)
-
-    fun getMuteElementTypes(entity: Entity): List<ElementType>
-
-    fun hasMuteElementType(entity: Entity, elementType: ElementType): Boolean
-
-}
\ No newline at end of file
