From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DeeChael <DeeChael@outlook.com>
Date: Tue, 19 Sep 2023 00:08:36 +0800
Subject: [PATCH] feat: protocol manager api


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index c36aca9bb5a0f8a67aa84a690a010445bd7712af..80afaca0f316edfdf65e1bb341c18843094bfce6 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -15,6 +15,7 @@ import java.util.UUID;
 import java.util.function.Consumer;
 import java.util.logging.Logger;
 
+import net.deechael.protocol.ProtocolManager;
 import net.deechael.elements.api.ElementManager;
 import net.deechael.skills.api.SkillManager;
 import org.bukkit.Warning.WarningState;
@@ -2767,6 +2768,10 @@ public final class Bukkit {
     public static SkillManager getSkillManager() {
         return server.getSkillManager();
     }
+
+    public static ProtocolManager getProtocolManager() {
+        return server.getProtocolManager();
+    }
     // Artist end
 
     @NotNull
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index b0904d838ef2d5690d5e0d1a26ea15895f480ff6..f9b9ac7edfa02a063d10f60da67552aec25546b1 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -15,6 +15,7 @@ import java.util.UUID;
 import java.util.function.Consumer;
 import java.util.logging.Logger;
 
+import net.deechael.protocol.ProtocolManager;
 import net.deechael.elements.api.ElementManager;
 import net.deechael.skills.api.SkillManager;
 import org.bukkit.Warning.WarningState;
@@ -2416,5 +2417,7 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     ElementManager getElementManager();
 
     SkillManager getSkillManager();
+
+    ProtocolManager getProtocolManager();
     // Artist end
 }
diff --git a/src/main/kotlin/net/deechael/protocol/ProtocolListener.kt b/src/main/kotlin/net/deechael/protocol/ProtocolListener.kt
new file mode 100644
index 0000000000000000000000000000000000000000..fcfba784b1bad6bf2e6a19fb74b0325dee1ea746
--- /dev/null
+++ b/src/main/kotlin/net/deechael/protocol/ProtocolListener.kt
@@ -0,0 +1,12 @@
+package net.deechael.protocol
+
+import net.deechael.protocol.packet.Packet
+import org.bukkit.entity.Player
+
+interface ProtocolListener {
+
+    fun sending(player: Player, packet: Packet): Packet?
+
+    fun receiving(player: Player, packet: Packet): Packet?
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/protocol/ProtocolManager.kt b/src/main/kotlin/net/deechael/protocol/ProtocolManager.kt
new file mode 100644
index 0000000000000000000000000000000000000000..9015cff777322137903153513ccc08ed1d2625d2
--- /dev/null
+++ b/src/main/kotlin/net/deechael/protocol/ProtocolManager.kt
@@ -0,0 +1,11 @@
+package net.deechael.protocol
+
+import org.bukkit.plugin.Plugin
+
+interface ProtocolManager {
+
+    fun registerListener(plugin: Plugin, listener: ProtocolListener)
+
+    fun unregisterAllListener(plugin: Plugin)
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/protocol/packet/Packet.kt b/src/main/kotlin/net/deechael/protocol/packet/Packet.kt
new file mode 100644
index 0000000000000000000000000000000000000000..3148b86d8446566e9e82684f7daeb22a43a666aa
--- /dev/null
+++ b/src/main/kotlin/net/deechael/protocol/packet/Packet.kt
@@ -0,0 +1,9 @@
+package net.deechael.protocol.packet
+
+interface Packet {
+
+    fun original(): Any
+
+    fun cast(): Any
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/protocol/packet/S2CSetActionBarPacket.kt b/src/main/kotlin/net/deechael/protocol/packet/S2CSetActionBarPacket.kt
new file mode 100644
index 0000000000000000000000000000000000000000..ae8a2af71c0e142b2c89e5ca0c4317b366492cfd
--- /dev/null
+++ b/src/main/kotlin/net/deechael/protocol/packet/S2CSetActionBarPacket.kt
@@ -0,0 +1,10 @@
+package net.deechael.protocol.packet
+
+import net.kyori.adventure.text.Component
+
+interface S2CSetActionBarPacket : Packet {
+
+    fun adventureContent(): Component
+
+    fun adventureContent(component: Component)
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/protocol/packet/S2CSetSubtitlePacket.kt b/src/main/kotlin/net/deechael/protocol/packet/S2CSetSubtitlePacket.kt
new file mode 100644
index 0000000000000000000000000000000000000000..f06d2b8a511348267caadf5130c2cfab99ce8b64
--- /dev/null
+++ b/src/main/kotlin/net/deechael/protocol/packet/S2CSetSubtitlePacket.kt
@@ -0,0 +1,10 @@
+package net.deechael.protocol.packet
+
+import net.kyori.adventure.text.Component
+
+interface S2CSetSubtitlePacket : Packet {
+
+    fun adventureContent(): Component
+
+    fun adventureContent(component: Component)
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/protocol/packet/S2CSetTitlePacket.kt b/src/main/kotlin/net/deechael/protocol/packet/S2CSetTitlePacket.kt
new file mode 100644
index 0000000000000000000000000000000000000000..fcdae2e9b0bea72378f0637b1c5a44b6706fba55
--- /dev/null
+++ b/src/main/kotlin/net/deechael/protocol/packet/S2CSetTitlePacket.kt
@@ -0,0 +1,10 @@
+package net.deechael.protocol.packet
+
+import net.kyori.adventure.text.Component
+
+interface S2CSetTitlePacket : Packet {
+
+    fun adventureContent(): Component
+
+    fun adventureContent(component: Component)
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/protocol/packet/S2CSystemChatPacket.kt b/src/main/kotlin/net/deechael/protocol/packet/S2CSystemChatPacket.kt
new file mode 100644
index 0000000000000000000000000000000000000000..af022c8e760a9a3a907ac169016deadf379d99f5
--- /dev/null
+++ b/src/main/kotlin/net/deechael/protocol/packet/S2CSystemChatPacket.kt
@@ -0,0 +1,11 @@
+package net.deechael.protocol.packet
+
+import net.kyori.adventure.text.Component
+
+interface S2CSystemChatPacket : Packet {
+
+    fun adventureContent(): Component
+
+    fun adventureContent(component: Component)
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/protocol/packet/S2CTabListPacket.kt b/src/main/kotlin/net/deechael/protocol/packet/S2CTabListPacket.kt
new file mode 100644
index 0000000000000000000000000000000000000000..6d1bbdb3204345f0fe40f4b856373b37a4b7a942
--- /dev/null
+++ b/src/main/kotlin/net/deechael/protocol/packet/S2CTabListPacket.kt
@@ -0,0 +1,15 @@
+package net.deechael.protocol.packet
+
+import net.kyori.adventure.text.Component
+
+interface S2CTabListPacket : Packet {
+
+    fun headerContent(): Component
+
+    fun headerContent(component: Component)
+
+    fun footerContent(): Component
+
+    fun footerContent(component: Component)
+
+}
\ No newline at end of file
