From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DeeChael <DeeChael@outlook.com>
Date: Thu, 31 Aug 2023 14:46:55 +0800
Subject: [PATCH] add skill api


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 0ee9a1287cfd6d72438b60fa88ecc58a1e6b856c..9bd371311e1ed6c39515d9c410b82b11198a830b 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -6,6 +6,7 @@ import java.util.Set;
 import java.util.UUID;
 
 import net.deechael.elements.api.application.ElementApplication;
+import net.deechael.skills.api.SkillContainer;
 import org.bukkit.FluidCollisionMode;
 import org.bukkit.Location;
 import org.bukkit.Material;
@@ -1195,5 +1196,10 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
     void setBodyYaw(float bodyYaw);
     // Paper end
 
+    // Artist start
     ElementApplication getElementApplication();
+
+    SkillContainer getSkillContainer();
+    // Artist end
+
 }
diff --git a/src/main/kotlin/net/deechael/skills/api/PassiveSkill.kt b/src/main/kotlin/net/deechael/skills/api/PassiveSkill.kt
new file mode 100644
index 0000000000000000000000000000000000000000..4ed61ba3d70f6cda4f9252d89a2cad365c8926b4
--- /dev/null
+++ b/src/main/kotlin/net/deechael/skills/api/PassiveSkill.kt
@@ -0,0 +1,5 @@
+package net.deechael.skills.api
+
+abstract class PassiveSkill(id: String, cooldown: Double, manaCost: Int) : Skill(id, cooldown, manaCost) {
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/skills/api/Skill.kt b/src/main/kotlin/net/deechael/skills/api/Skill.kt
new file mode 100644
index 0000000000000000000000000000000000000000..82f667d907c2d4d89db16ae351fb5d7eab3449a8
--- /dev/null
+++ b/src/main/kotlin/net/deechael/skills/api/Skill.kt
@@ -0,0 +1,26 @@
+package net.deechael.skills.api
+
+import org.bukkit.entity.LivingEntity
+import org.bukkit.inventory.ItemStack
+
+abstract class Skill(
+    private val id: String,
+    private val cooldown: Double,
+    private val manaCost: Int
+) {
+
+    fun getId(): String {
+        return this.id
+    }
+
+    fun getCooldown(): Double {
+        return this.cooldown
+    }
+
+    fun getManaCost(): Int {
+        return this.manaCost
+    }
+
+    abstract fun trigger(commander: LivingEntity, skillItem: ItemStack?)
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/skills/api/SkillContainer.kt b/src/main/kotlin/net/deechael/skills/api/SkillContainer.kt
new file mode 100644
index 0000000000000000000000000000000000000000..4690c7e6ecff285136805a9105df27cf36ff714d
--- /dev/null
+++ b/src/main/kotlin/net/deechael/skills/api/SkillContainer.kt
@@ -0,0 +1,17 @@
+package net.deechael.skills.api
+
+interface SkillContainer {
+
+    fun getPassiveSkill(): PassiveSkill?
+
+    fun setPassiveSkill(passiveSkill: PassiveSkill?)
+
+    fun hasPassiveSkill(): Boolean
+
+    fun getSkill(slot: SkillSlot): Skill?
+
+    fun setSkill(slot: SkillSlot, skill: Skill?)
+
+    fun hasSkill(slot: SkillSlot): Boolean
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/skills/api/SkillManager.kt b/src/main/kotlin/net/deechael/skills/api/SkillManager.kt
new file mode 100644
index 0000000000000000000000000000000000000000..1d7828949c014fa11fa224c5a38878ba3cdf4ce4
--- /dev/null
+++ b/src/main/kotlin/net/deechael/skills/api/SkillManager.kt
@@ -0,0 +1,13 @@
+package net.deechael.skills.api
+
+import org.bukkit.NamespacedKey
+
+interface SkillManager {
+
+    fun registerSkill(key: NamespacedKey, skill: Skill)
+
+    fun getSkill(key: NamespacedKey)
+
+    fun getSkills(): Map<NamespacedKey, Skill>
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/skills/api/SkillSlot.kt b/src/main/kotlin/net/deechael/skills/api/SkillSlot.kt
new file mode 100644
index 0000000000000000000000000000000000000000..ce350c6281a4e224ea1b5bc4ca1bf0f0a06e61bb
--- /dev/null
+++ b/src/main/kotlin/net/deechael/skills/api/SkillSlot.kt
@@ -0,0 +1,9 @@
+package net.deechael.skills.api
+
+enum class SkillSlot {
+
+    PRIMARY,
+    SECONDARY,
+    THIRDARY
+
+}
\ No newline at end of file
