From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DeeChael <DeeChael@outlook.com>
Date: Thu, 21 Sep 2023 02:01:07 +0800
Subject: [PATCH] feat: gui api


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 9ebdcd00b9cfb37efd7abdc169f0eac6ff7e430e..4528bdcebb446082b39a75c430a9750d408b192e 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -16,6 +16,7 @@ import java.util.function.Consumer;
 import java.util.logging.Logger;
 
 import net.deechael.command.ArtistCommandManager;
+import net.deechael.gui.GuiManager;
 import net.deechael.nbt.NbtManager;
 import net.deechael.protocol.ProtocolManager;
 import net.deechael.elements.api.ElementManager;
@@ -2782,6 +2783,10 @@ public final class Bukkit {
     public static NbtManager getNbtManager() {
         return server.getNbtManager();
     }
+
+    public static GuiManager getGuiManager() {
+        return server.getGuiManager();
+    }
     // Artist end
 
     @NotNull
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 1e1ec69d398a282d440d670308056794fa16d354..1f606841d4a764ef797e435282467ad65eede186 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -16,6 +16,7 @@ import java.util.function.Consumer;
 import java.util.logging.Logger;
 
 import net.deechael.command.ArtistCommandManager;
+import net.deechael.gui.GuiManager;
 import net.deechael.nbt.NbtManager;
 import net.deechael.protocol.ProtocolManager;
 import net.deechael.elements.api.ElementManager;
@@ -2425,5 +2426,7 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     ArtistCommandManager getArtistCommandManager();
 
     NbtManager getNbtManager();
+
+    GuiManager getGuiManager();
     // Artist end
 }
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index f9dd00210c1762a40259f823aeb8d8a5ddc78e3e..c823ed9d6491f355537ac9bfde341dadd50c58c4 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -7,6 +7,8 @@ import java.time.Instant;
 import java.util.Collection;
 import java.util.Date;
 import java.util.Map;
+
+import net.deechael.gui.GuiAudience;
 import org.bukkit.BanEntry;
 import org.bukkit.DyeColor;
 import org.bukkit.Effect;
@@ -51,7 +53,7 @@ import org.jetbrains.annotations.Nullable;
 /**
  * Represents a player, connected or not
  */
-public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginMessageRecipient, net.kyori.adventure.identity.Identified, net.kyori.adventure.bossbar.BossBarViewer, com.destroystokyo.paper.network.NetworkClient { // Paper
+public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginMessageRecipient, net.kyori.adventure.identity.Identified, net.kyori.adventure.bossbar.BossBarViewer, com.destroystokyo.paper.network.NetworkClient, GuiAudience { // Paper
 
     // Paper start
     @Override
diff --git a/src/main/kotlin/net/deechael/gui/Clicker.kt b/src/main/kotlin/net/deechael/gui/Clicker.kt
new file mode 100644
index 0000000000000000000000000000000000000000..71832eaf33fbe5683c77e06bd983e4ce7f9b6a73
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/Clicker.kt
@@ -0,0 +1,9 @@
+package net.deechael.gui
+
+import org.bukkit.inventory.ItemStack
+
+fun interface Clicker {
+
+    fun click(context: SlotContext, cursor: ItemStack?)
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/Closer.kt b/src/main/kotlin/net/deechael/gui/Closer.kt
new file mode 100644
index 0000000000000000000000000000000000000000..478bcc4552115af9722fbb5ecec226e7e4830f7a
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/Closer.kt
@@ -0,0 +1,9 @@
+package net.deechael.gui
+
+import net.deechael.gui.Scene
+
+fun interface Closer {
+
+    fun close(scene: Scene)
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/Gui.kt b/src/main/kotlin/net/deechael/gui/Gui.kt
new file mode 100644
index 0000000000000000000000000000000000000000..3ab2c1fe7ac9df2a52250e136f217a59021721fb
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/Gui.kt
@@ -0,0 +1,11 @@
+package net.deechael.gui
+
+interface Gui {
+
+    fun audiences(): List<GuiAudience>
+
+    fun render(audience: GuiAudience): GuiRenderer
+
+    fun scene(audience: GuiAudience): Scene?
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/GuiAudience.kt b/src/main/kotlin/net/deechael/gui/GuiAudience.kt
new file mode 100644
index 0000000000000000000000000000000000000000..81423577ed949463a214a4938bbc3d712558160b
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/GuiAudience.kt
@@ -0,0 +1,7 @@
+package net.deechael.gui
+
+interface GuiAudience {
+
+    fun getGuiContext(): GuiContext
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/GuiContext.kt b/src/main/kotlin/net/deechael/gui/GuiContext.kt
new file mode 100644
index 0000000000000000000000000000000000000000..9441a650b4330016c58998547cf5dd178f7762de
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/GuiContext.kt
@@ -0,0 +1,11 @@
+package net.deechael.gui
+
+interface GuiContext {
+
+    fun getScene(): Scene?
+
+    fun setScene(scene: Scene)
+
+    fun removeScene()
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/GuiManager.kt b/src/main/kotlin/net/deechael/gui/GuiManager.kt
new file mode 100644
index 0000000000000000000000000000000000000000..4607403cff9d2ceae04794699e20bfbb7a7dd144
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/GuiManager.kt
@@ -0,0 +1,9 @@
+package net.deechael.gui
+
+import org.bukkit.plugin.Plugin
+
+interface GuiManager {
+
+    fun <T: Gui> createGui(plugin: Plugin, type: GuiType) : T
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/GuiRenderer.kt b/src/main/kotlin/net/deechael/gui/GuiRenderer.kt
new file mode 100644
index 0000000000000000000000000000000000000000..bed86be0167777767feb66aae2926ede9c3eb640
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/GuiRenderer.kt
@@ -0,0 +1,17 @@
+package net.deechael.gui
+
+import net.kyori.adventure.text.Component
+
+interface GuiRenderer {
+
+    fun gui(): Gui
+
+    fun audience(): GuiAudience
+
+    fun title(title: Component?): GuiRenderer
+
+    fun close(close: Closer): GuiRenderer
+
+    fun render(): Scene
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/GuiType.kt b/src/main/kotlin/net/deechael/gui/GuiType.kt
new file mode 100644
index 0000000000000000000000000000000000000000..06ea3040a439ea47019a2c7819f05c5103019c91
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/GuiType.kt
@@ -0,0 +1,23 @@
+package net.deechael.gui
+
+import net.deechael.gui.anvil.AnvilGui
+import net.deechael.gui.normal.NormalGui
+import net.deechael.gui.pageable.PageableGui
+
+enum class GuiType(val guiClass: Class<out Gui>) {
+
+    NORMAL_1X9(NormalGui::class.java),
+    NORMAL_2X9(NormalGui::class.java),
+    NORMAL_3X9(NormalGui::class.java),
+    NORMAL_4X9(NormalGui::class.java),
+    NORMAL_5X9(NormalGui::class.java),
+    NORMAL_6X9(NormalGui::class.java),
+    PAGEABLE_1X9(PageableGui::class.java),
+    PAGEABLE_2X9(PageableGui::class.java),
+    PAGEABLE_3X9(PageableGui::class.java),
+    PAGEABLE_4X9(PageableGui::class.java),
+    PAGEABLE_5X9(PageableGui::class.java),
+    PAGEABLE_6X9(PageableGui::class.java),
+    ANVIL(AnvilGui::class.java)
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/Imager.kt b/src/main/kotlin/net/deechael/gui/Imager.kt
new file mode 100644
index 0000000000000000000000000000000000000000..0f5a85a48112ec09cb99846dd2f67802a255a825
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/Imager.kt
@@ -0,0 +1,9 @@
+package net.deechael.gui
+
+import org.bukkit.inventory.ItemStack
+
+fun interface Imager {
+
+    fun render(context: SlotContext): ItemStack?
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/Scene.kt b/src/main/kotlin/net/deechael/gui/Scene.kt
new file mode 100644
index 0000000000000000000000000000000000000000..d3535dba2efdf7d8cc1147b5a82f4526b675d4bf
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/Scene.kt
@@ -0,0 +1,15 @@
+package net.deechael.gui
+
+import org.bukkit.inventory.Inventory
+
+interface Scene {
+
+    fun gui(): Gui
+
+    fun audience(): GuiAudience
+
+    fun refresh(keepPrevious: Boolean): GuiRenderer
+
+    fun asBukkit(): Inventory
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/Slot.kt b/src/main/kotlin/net/deechael/gui/Slot.kt
new file mode 100644
index 0000000000000000000000000000000000000000..1ac2c7bd2f2992014f33ed508de3ca996a36b425
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/Slot.kt
@@ -0,0 +1,9 @@
+package net.deechael.gui
+
+interface Slot {
+
+    fun image(image: Imager)
+
+    fun click(click: Clicker)
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/SlotContext.kt b/src/main/kotlin/net/deechael/gui/SlotContext.kt
new file mode 100644
index 0000000000000000000000000000000000000000..a64bf94b0c0e136048c634feff4a56c4c7423415
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/SlotContext.kt
@@ -0,0 +1,14 @@
+package net.deechael.gui
+
+import org.bukkit.event.inventory.ClickType
+import org.bukkit.event.inventory.InventoryAction
+
+interface SlotContext {
+
+    fun scene(): Scene
+
+    fun clickType(): ClickType?
+
+    fun action(): InventoryAction?
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/anvil/AnvilGui.kt b/src/main/kotlin/net/deechael/gui/anvil/AnvilGui.kt
new file mode 100644
index 0000000000000000000000000000000000000000..52fce0b190aa67cb42ed5ecfa3a48f3718f00cba
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/anvil/AnvilGui.kt
@@ -0,0 +1,12 @@
+package net.deechael.gui.anvil
+
+import net.deechael.gui.Gui
+import net.deechael.gui.GuiAudience
+import net.deechael.gui.GuiRenderer
+import net.deechael.gui.Scene
+
+interface AnvilGui : Gui {
+
+    override fun render(audience: GuiAudience): AnvilGuiRenderer
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/anvil/AnvilGuiRenderer.kt b/src/main/kotlin/net/deechael/gui/anvil/AnvilGuiRenderer.kt
new file mode 100644
index 0000000000000000000000000000000000000000..aadbfa936528cbcaef0ee75da87ed5b0c391c97b
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/anvil/AnvilGuiRenderer.kt
@@ -0,0 +1,19 @@
+package net.deechael.gui.anvil
+
+import net.deechael.gui.GuiRenderer
+import net.deechael.gui.Slot
+
+interface AnvilGuiRenderer : GuiRenderer {
+
+    override fun gui(): AnvilGui
+
+    fun leftInput(builder: Slot.() -> Unit): AnvilGuiRenderer
+
+    fun rightInput(builder: Slot.() -> Unit): AnvilGuiRenderer
+
+    fun result(builder: Slot.() -> Unit): AnvilGuiRenderer
+
+    fun resulter(builder: AnvilResulter): AnvilGuiRenderer
+
+    override fun render(): AnvilGuiScene
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/anvil/AnvilGuiScene.kt b/src/main/kotlin/net/deechael/gui/anvil/AnvilGuiScene.kt
new file mode 100644
index 0000000000000000000000000000000000000000..411d82fe78f8ec1c2289a344eb6ef1913c5ff014
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/anvil/AnvilGuiScene.kt
@@ -0,0 +1,12 @@
+package net.deechael.gui.anvil
+
+import net.deechael.gui.Scene
+import org.bukkit.inventory.ItemStack
+
+interface AnvilGuiScene: Scene {
+
+    var text: String
+
+    fun item(slot: AnvilSlot): ItemStack?
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/anvil/AnvilResulter.kt b/src/main/kotlin/net/deechael/gui/anvil/AnvilResulter.kt
new file mode 100644
index 0000000000000000000000000000000000000000..effa773539cde2b8c2cacdf52024f935841d8e65
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/anvil/AnvilResulter.kt
@@ -0,0 +1,9 @@
+package net.deechael.gui.anvil
+
+import org.bukkit.inventory.ItemStack
+
+fun interface AnvilResulter {
+
+    fun result(context: AnvilSlotContext): ItemStack?
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/anvil/AnvilSlot.kt b/src/main/kotlin/net/deechael/gui/anvil/AnvilSlot.kt
new file mode 100644
index 0000000000000000000000000000000000000000..84bed336c867f43539a2dd10f21fe503e924b78c
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/anvil/AnvilSlot.kt
@@ -0,0 +1,19 @@
+package net.deechael.gui.anvil
+
+enum class AnvilSlot(val slotIndex: Int) {
+
+    LEFT_INPUT(0),
+    RIGHT_INPUT(1),
+    RESULT(2);
+
+    companion object {
+        fun from(index: Int) : AnvilSlot {
+            return when (index) {
+                2 -> RESULT
+                1 -> RIGHT_INPUT
+                else -> LEFT_INPUT
+            }
+        }
+    }
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/anvil/AnvilSlotContext.kt b/src/main/kotlin/net/deechael/gui/anvil/AnvilSlotContext.kt
new file mode 100644
index 0000000000000000000000000000000000000000..f031dc7e45cb81b5ce35745d22a85f97ddf1a28d
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/anvil/AnvilSlotContext.kt
@@ -0,0 +1,17 @@
+package net.deechael.gui.anvil
+
+import net.deechael.gui.SlotContext
+import org.bukkit.inventory.ItemStack
+
+interface AnvilSlotContext : SlotContext {
+
+    fun slot(): AnvilSlot
+
+    fun leftInput(): ItemStack?
+
+    fun rightInput(): ItemStack?
+
+    fun result(): ItemStack?
+
+    fun text(): String
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/normal/NormalGui.kt b/src/main/kotlin/net/deechael/gui/normal/NormalGui.kt
new file mode 100644
index 0000000000000000000000000000000000000000..159cef322a1224837e81459c441f8d334e0ff6e6
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/normal/NormalGui.kt
@@ -0,0 +1,12 @@
+package net.deechael.gui.normal
+
+import net.deechael.gui.Gui
+import net.deechael.gui.GuiAudience
+
+interface NormalGui : Gui {
+
+    fun lines(): Int
+
+    override fun render(audience: GuiAudience): NormalGuiRenderer
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/normal/NormalGuiRenderer.kt b/src/main/kotlin/net/deechael/gui/normal/NormalGuiRenderer.kt
new file mode 100644
index 0000000000000000000000000000000000000000..e45849c568f21291f5398b80b2f6737eb4bedca6
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/normal/NormalGuiRenderer.kt
@@ -0,0 +1,15 @@
+package net.deechael.gui.normal
+
+import net.deechael.gui.GuiRenderer
+import net.deechael.gui.Scene
+import net.deechael.gui.Slot
+
+interface NormalGuiRenderer: GuiRenderer {
+
+    override fun gui(): NormalGui
+
+    fun slot(raw: Int, builder: Slot.() -> Unit)
+
+    override fun render(): NormalGuiScene
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/normal/NormalGuiScene.kt b/src/main/kotlin/net/deechael/gui/normal/NormalGuiScene.kt
new file mode 100644
index 0000000000000000000000000000000000000000..a89240d49ef76c89a3f0fa25ce61d1b089e4a6e2
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/normal/NormalGuiScene.kt
@@ -0,0 +1,12 @@
+package net.deechael.gui.normal
+
+import net.deechael.gui.Scene
+import org.bukkit.inventory.ItemStack
+
+interface NormalGuiScene : Scene {
+
+    override fun refresh(keepPrevious: Boolean): NormalGuiRenderer
+
+    fun item(raw: Int): ItemStack?
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/normal/NormalSlotContext.kt b/src/main/kotlin/net/deechael/gui/normal/NormalSlotContext.kt
new file mode 100644
index 0000000000000000000000000000000000000000..141b53d2b114af5045faf75e6ba33f37815ce1f9
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/normal/NormalSlotContext.kt
@@ -0,0 +1,9 @@
+package net.deechael.gui.normal
+
+import net.deechael.gui.SlotContext
+
+interface NormalSlotContext : SlotContext {
+
+    fun slot(): Int
+
+}
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/gui/pageable/PageableGui.kt b/src/main/kotlin/net/deechael/gui/pageable/PageableGui.kt
new file mode 100644
index 0000000000000000000000000000000000000000..1c350ed68ff8d51688bdc82d26cda8a6463a7447
--- /dev/null
+++ b/src/main/kotlin/net/deechael/gui/pageable/PageableGui.kt
@@ -0,0 +1,6 @@
+package net.deechael.gui.pageable
+
+import net.deechael.gui.Gui
+
+interface PageableGui : Gui {
+}
\ No newline at end of file
