From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DeeChael <DeeChael@outlook.com>
Date: Tue, 19 Sep 2023 02:03:08 +0800
Subject: [PATCH] fix: command


diff --git a/src/main/kotlin/net/deechael/command/ArgumentType.kt b/src/main/kotlin/net/deechael/command/ArgumentType.kt
index 7bfc3cd49d27acb7326a4738bdac49fccddd2b64..103f3a5121dee68c4f897338862181e57f9c8afd 100644
--- a/src/main/kotlin/net/deechael/command/ArgumentType.kt
+++ b/src/main/kotlin/net/deechael/command/ArgumentType.kt
@@ -2,8 +2,6 @@ package net.deechael.command
 
 interface ArgumentType<T> {
 
-    fun get(context: CommandContext, name: String): T {
-        return context.get(name, this)
-    }
+    fun get(context: CommandContext, name: String): T
 
 }
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/command/CommandContext.kt b/src/main/kotlin/net/deechael/command/CommandContext.kt
index d8fe8a681110207eef989a08f9dd57f73c5cee1e..83501138498734602f053db121d4a5923e372a62 100644
--- a/src/main/kotlin/net/deechael/command/CommandContext.kt
+++ b/src/main/kotlin/net/deechael/command/CommandContext.kt
@@ -4,8 +4,6 @@ import org.bukkit.command.CommandSender
 
 interface CommandContext {
 
-    fun <T, A : ArgumentType<T>> get(name: String, argumentType: A): T
-
     fun getSender(): CommandSender
 
 }
\ No newline at end of file
