From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DeeChael <DeeChael@outlook.com>
Date: Fri, 1 Sep 2023 00:45:33 +0800
Subject: [PATCH] fix field signature problem


diff --git a/src/main/kotlin/net/deechael/artist/skill/trigger/AbstractTriggerEnvironment.kt b/src/main/kotlin/net/deechael/artist/skill/trigger/AbstractTriggerEnvironment.kt
index b2d25cddd894e292078770460b025ec9a2ca548f..72cf9f1d69685cd7c90b756633ba6c4aded0e05d 100644
--- a/src/main/kotlin/net/deechael/artist/skill/trigger/AbstractTriggerEnvironment.kt
+++ b/src/main/kotlin/net/deechael/artist/skill/trigger/AbstractTriggerEnvironment.kt
@@ -4,7 +4,7 @@ import net.deechael.skills.api.trigger.TriggerEnvironment
 import net.minecraft.world.entity.LivingEntity
 import org.bukkit.inventory.ItemStack
 
-abstract class AbstractTriggerEnvironment(val target: LivingEntity, val itemStack: ItemStack?): TriggerEnvironment {
+abstract class AbstractTriggerEnvironment(private val target: LivingEntity, private val itemStack: ItemStack?): TriggerEnvironment {
 
     override fun getTarget(): org.bukkit.entity.LivingEntity {
         return this.target.bukkitLivingEntity
