From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DeeChael <DeeChael@outlook.com>
Date: Fri, 1 Sep 2023 00:58:57 +0800
Subject: [PATCH] fix ender dragon part not implements method


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragonPart.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragonPart.java
index a407e802164603c74ec1f4ad3c98bc50398954bc..2a01c2a19fcd0a0b00857dc6f9d738c72223e79c 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragonPart.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragonPart.java
@@ -1,5 +1,6 @@
 package org.bukkit.craftbukkit.entity;
 
+import net.deechael.elements.api.application.source.ApplicationSource;
 import org.bukkit.craftbukkit.CraftServer;
 import org.bukkit.entity.EnderDragon;
 import org.bukkit.entity.EnderDragonPart;
@@ -69,4 +70,9 @@ public class CraftEnderDragonPart extends CraftComplexPart implements EnderDrago
     public void resetMaxHealth() {
         this.getParent().resetMaxHealth();
     }
+
+    @Override
+    public void elementalDamage(double amount, ApplicationSource source) {
+        this.getParent().elementalDamage(amount, source);
+    }
 }
