From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DeeChael <DeeChael@outlook.com>
Date: Thu, 21 Sep 2023 22:54:15 +0800
Subject: [PATCH] fix: more


diff --git a/src/main/kotlin/net/deechael/artist/gui/ArtistGuiManager.kt b/src/main/kotlin/net/deechael/artist/gui/ArtistGuiManager.kt
index 01707e6e61ec615257980a826b4514a96e998d9a..597d3e434e835ce9840bd0e1d431996dd29d5232 100644
--- a/src/main/kotlin/net/deechael/artist/gui/ArtistGuiManager.kt
+++ b/src/main/kotlin/net/deechael/artist/gui/ArtistGuiManager.kt
@@ -20,6 +20,7 @@ object ArtistGuiManager : GuiManager {
     override fun <T : Gui> createGui(plugin: Plugin, type: GuiType): T {
         val guiImpl = interfaceToImplements[type.guiClass]!!
         val construct = guiImpl.getDeclaredConstructor(Plugin::class.java, GuiType::class.java)
+        construct.isAccessible = true
         return construct.newInstance(plugin, type) as T
     }
 
diff --git a/src/main/kotlin/net/deechael/artist/gui/anvil/ArtistAnvilGui.kt b/src/main/kotlin/net/deechael/artist/gui/anvil/ArtistAnvilGui.kt
index fc241cb57235ee4636761dec91076337b113f671..8567bd4f8e0abed35574f868d60f4989e604e2aa 100644
--- a/src/main/kotlin/net/deechael/artist/gui/anvil/ArtistAnvilGui.kt
+++ b/src/main/kotlin/net/deechael/artist/gui/anvil/ArtistAnvilGui.kt
@@ -2,14 +2,15 @@ package net.deechael.artist.gui.anvil
 
 import net.deechael.artist.gui.ArtistGui
 import net.deechael.gui.GuiAudience
+import net.deechael.gui.GuiType
 import net.deechael.gui.anvil.AnvilGui
 import net.deechael.gui.anvil.AnvilGuiRenderer
 import org.bukkit.plugin.Plugin
 
-class ArtistAnvilGui(plugin: Plugin) : ArtistGui(plugin), AnvilGui {
+class ArtistAnvilGui(plugin: Plugin, type: GuiType) : ArtistGui(plugin), AnvilGui {
 
     override fun render(audience: GuiAudience): AnvilGuiRenderer {
-        TODO("Not yet implemented")
+        return ArtistAnvilGuiRenderer(this.plugin, this, audience, null)
     }
 
 }
\ No newline at end of file
diff --git a/src/main/kotlin/net/deechael/artist/gui/anvil/ArtistAnvilGuiRenderer.kt b/src/main/kotlin/net/deechael/artist/gui/anvil/ArtistAnvilGuiRenderer.kt
index 03320c45ab08b0cb06c925668c8268eb1ccb6cb2..9a2add55a5d86b02ba29aa60a080c1d7ade43f6a 100644
--- a/src/main/kotlin/net/deechael/artist/gui/anvil/ArtistAnvilGuiRenderer.kt
+++ b/src/main/kotlin/net/deechael/artist/gui/anvil/ArtistAnvilGuiRenderer.kt
@@ -16,8 +16,10 @@ class ArtistAnvilGuiRenderer(
     plugin: Plugin,
     gui: Gui,
     audience: GuiAudience,
-    private val artistAnvilGuiScene: ArtistAnvilGuiScene?) : ArtistGuiRenderer(plugin, gui, audience), AnvilGuiRenderer {
+    private val artistAnvilGuiScene: ArtistAnvilGuiScene?
+) : ArtistGuiRenderer(plugin, gui, audience), AnvilGuiRenderer {
 
+    private var text: String? = null
     private var leftInput: ArtistSlot? = null
     private var rightInput: ArtistSlot? = null
     private var result: ArtistSlot? = null
@@ -47,9 +49,14 @@ class ArtistAnvilGuiRenderer(
         return this
     }
 
+    override fun text(text: String): AnvilGuiRenderer {
+        this.text = text
+        return this;
+    }
+
     override fun render(): AnvilGuiScene {
         val scene = this.artistAnvilGuiScene ?: ArtistAnvilGuiScene(this.title, this.gui(), this.audience())
-        scene.rerender(this.title, this.close, this.leftInput, this.rightInput, this.result, this.resulter)
+        scene.rerender(this.title, this.close, this.text, this.leftInput, this.rightInput, this.result, this.resulter)
         return scene
     }
 
diff --git a/src/main/kotlin/net/deechael/artist/gui/anvil/ArtistAnvilGuiScene.kt b/src/main/kotlin/net/deechael/artist/gui/anvil/ArtistAnvilGuiScene.kt
index 8d249fc9a8c13690c12973dfba6509f2e5781660..50249a5e5d240ca15a42a61a3396cb8168b73b8b 100644
--- a/src/main/kotlin/net/deechael/artist/gui/anvil/ArtistAnvilGuiScene.kt
+++ b/src/main/kotlin/net/deechael/artist/gui/anvil/ArtistAnvilGuiScene.kt
@@ -80,7 +80,7 @@ class ArtistAnvilGuiScene(title: Component?, gui: Gui, audience: GuiAudience) :
         return ArtistAnvilGuiRenderer(this.gui().plugin, this.gui, this.audience, this)
     }
 
-    fun rerender(title: Component?, closer: Closer, leftInput: ArtistSlot?, rightInput: ArtistSlot?, result: ArtistSlot?, resulter: AnvilResulter?) {
+    fun rerender(title: Component?, closer: Closer, text: String?, leftInput: ArtistSlot?, rightInput: ArtistSlot?, result: ArtistSlot?, resulter: AnvilResulter?) {
         if (title != null) {
             val serverPlayer = (this.audience as CraftPlayer).handle
             val titlePacket = ClientboundOpenScreenPacket(
@@ -90,6 +90,9 @@ class ArtistAnvilGuiScene(title: Component?, gui: Gui, audience: GuiAudience) :
             )
             serverPlayer.connection.send(titlePacket)
         }
+        if (text != null) {
+            this.fakeAnvil.setText(text)
+        }
         this.closer = closer
         this.leftInput = leftInput
         this.rightInput = rightInput
diff --git a/src/main/kotlin/net/deechael/artist/gui/normal/ArtistNormalGuiRenderer.kt b/src/main/kotlin/net/deechael/artist/gui/normal/ArtistNormalGuiRenderer.kt
index 9d1cf2f5a117036ca932c62e645da08b2dbd4a6c..4d3e48249bb7227ed43445579cf06fd994c63afc 100644
--- a/src/main/kotlin/net/deechael/artist/gui/normal/ArtistNormalGuiRenderer.kt
+++ b/src/main/kotlin/net/deechael/artist/gui/normal/ArtistNormalGuiRenderer.kt
@@ -26,8 +26,9 @@ class ArtistNormalGuiRenderer(
         return super.gui() as NormalGui
     }
 
-    override fun slot(raw: Int, builder: Slot.() -> Unit) {
+    override fun slot(raw: Int, builder: Slot.() -> Unit): NormalGuiRenderer {
         this.slots[raw] = ArtistSlot().apply(builder)
+        return this
     }
 
     override fun render(): NormalGuiScene {
